vllm_gaudi/utils.py

```diff
@@ -29,6 +29,38 @@ def hpu_backend_string():
     return backend_string
 
 
+def async_h2d_copy(source, dest_tensor=None, dtype=None, device='hpu'):
+    """
+    Asynchronously transfer data from host to device.
+
+    Args:
+        source: CPU tensor or raw data to transfer
+        dest_tensor: Optional pre-allocated destination tensor
+        dtype: Required if source is raw data
+        device: Target device
+
+    Returns:
+        torch.Tensor on target device
+    """
+    if isinstance(source, torch.Tensor):
+        if dest_tensor is not None:
+            # Copy into pre-allocated destination tensor
+            return dest_tensor.copy_(source, non_blocking=True)
+        else:
+            # Create new device tensor and copy
+            assert source.device.type == 'cpu', \
+                "Source tensor must be on CPU for asynchronous transfer"
+            target = torch.empty_like(source, device=device)
+            return target.copy_(source, non_blocking=True)
+    else:
+        # Create tensor from data and transfer to device
+        if dtype is None:
+            raise ValueError(
+                "dtype must be specified when source is not a tensor")
+        cpu_tensor = torch.tensor(source, dtype=dtype, device='cpu')
+        return cpu_tensor.to(device, non_blocking=True)
```

Comment:
We could utilize a guard clause and reduce the number of if/else statements here and decrease the indents
```suggestion
    if isinstance(source, torch.Tensor):
        if dest_tensor is not None:
            # Copy into pre-allocated destination tensor
            return dest_tensor.copy_(source, non_blocking=True)
        # Create new device tensor and copy
        assert source.device.type == 'cpu', \
            "Source tensor must be on CPU for asynchronous transfer"
        target = torch.empty_like(source, device=device)
        return target.copy_(source, non_blocking=True)
    # Create tensor from data and transfer to device
    if dtype is None:
        raise ValueError(
            "dtype must be specified when source is not a tensor")
    cpu_tensor = torch.tensor(source, dtype=dtype, device='cpu')
    return cpu_tensor.to(device, non_blocking=True)
```